data_steward/analytics/cdr_ops/combined.py

```diff
@@ -0,0 +1,446 @@
+# report discrepancies to EHR Ops.
+#
+# ### Implementation notes
+# For each associated date/timestamp pair in the dataset the script that follows
+# assembles a universal subquery which can be applied to all pertinent tables and
+# whose results can be compiled together into a single result set that provides
+# relevant troubleshooting information. For legibility, a maximum of 10 problem
+# rows is returned for each table.
+#
+# As an example, the subquery for the `observation` table should look something
+# like this:
+#
+# ```sql
+# SELECT
+#  "observation"          AS table_name
+# ,observation_id         AS row_id
+# ,"observation_date"     AS date_field
+# ,observation_date       AS date_value
+# ,"observation_datetime" AS timestamp_field
+# ,observation_datetime   AS timestamp_value
+# FROM `DATASET_ID.observation`
+# WHERE observation_date<>DATE(observation_datetime)
+# LIMIT 10
+# ```
+#
+
+query = f'''
+DECLARE query DEFAULT (
+    WITH 
+
+     field_comparison AS
+     (SELECT 
+       d.table_schema AS table_schema
+      ,d.table_name   AS table_name
+      ,pk.column_name AS key_field
+      ,d.column_name  AS date_field
+      ,ts.column_name AS timestamp_field
+      FROM `{DATASET_ID}.INFORMATION_SCHEMA.COLUMNS` d
+       JOIN `{DATASET_ID}.INFORMATION_SCHEMA.COLUMNS` ts
+        ON d.table_name = ts.table_name
+           AND STARTS_WITH(ts.column_name, d.column_name)
+           AND d.data_type = 'DATE'
+           AND ts.data_type = 'TIMESTAMP'
+       JOIN `{DATASET_ID}.INFORMATION_SCHEMA.COLUMNS` pk
+        ON ts.table_name = pk.table_name
+           AND pk.ordinal_position=1)
+
+    SELECT
+     STRING_AGG(
+      '(SELECT ' 
+         || '"' || table_name        || '" AS table_name '
+         || ',' || key_field         || '  AS row_id '
+         || ',"' || date_field       || '" AS date_field '
+         || ',' || date_field        || '  AS date_value '
+         || ',"' || timestamp_field  || '" AS timestamp_field '
+         || ',' || timestamp_field   || '  AS timestamp_value '
+         || 'FROM `' || table_schema || '.' || table_name || '` '
+         || 'WHERE ' || date_field || '<>' || 'DATE(' || timestamp_field || ') '
+         || ' LIMIT 10 )'
+     ,' UNION ALL ')
+    FROM field_comparison
+);
+EXECUTE IMMEDIATE query;
+'''
+execute(query)
+
+# ---
+# # Manual Review
+# The following evaluation metrics require manual/visual review.
+
+# ## Verify dataset properties
+# The dataset metadata must have
+# 1. a description indicating the version of the pipeline run and the input datasets
+# 2. labels `phase`, `release_tag`, `de_identified` with appropriate values
+#
+# In the future this should try to find the dataset using labels supplied via notebook
+# parameters (e.g. `combined`, `release_tag=2021q3r1`, `phase=clean`). Currently
+# there is no way to specify `combined`.
+
+dataset = CLIENT.get_dataset(dataset_ref=f'{PROJECT_ID}.{DATASET_ID}')
+
+print(dataset.labels)
```

Comment:
This is much cleaner than querying for the labels.  It was a nice suggestion @jp3477 .  Is it feasible to just build the label checks in right here?  That way, no one has to think about this too deeply.
```
try:
    phase = dataset.labels['phase']
    deidentified = dataset.labels['de_identified']
    release_tag = dataset.labels['release_tag']
except KeyError as ke:
    print(f'dataset label validation failed because keys were missing entirely')
    print(ke.message)
else:
    invalid_labels = []
    if phase != 'clean':
        invalid_labels.append('phase')
    elif deidentified != 'false':
        invalid_labels.append('de_identified')
    elif !DATASET_ID.contains(release_tag):
        invalid_labels.append('release_tag')

    if invalid_labels:
        print(f'`{", ".join(invalid_labels)}` label(s) is not set to the expected value(s).  did a failure occur in the combined process?')
    else:
        print("labels pass basic inspection")
```